lib/indexer/indexer.js

```diff
@@ -0,0 +1,689 @@
+  /**
+   * Create an indexer.
+   * @constructor
+   * @param {String} module
+   * @param {Object} options
+   */
+
+  constructor(module, options) {
+    super();
+
+    assert(typeof module === 'string');
+    assert(module.length > 0);
+
+    this.options = new IndexOptions(module, options);
+
+    this.network = this.options.network;
+    this.logger = this.options.logger.context(`${module}indexer`);
+    this.blocks = this.options.blocks;
+    this.chain = this.options.chain;
+
+    this.db = null;
+    this.batch = null;
+    this.syncing = false;
+    this.height = 0;
+  }
+
+  /**
+   * Start a new batch write.
+   * @returns {Batch}
+   */
+
+  start() {
+    assert(this.batch === null, 'Already started.');
+    this.batch = this.db.batch();
+    return this.batch;
+  }
+
+  /**
+   * Put key and value to the current batch.
+   * @param {String} key
+   * @param {Buffer} value
+   */
+
+  put(key, value) {
+    this.batch.put(key, value);
+  }
+
+  /**
+   * Delete key from the current batch.
+   * @param {String} key
+   */
+
+  del(key) {
+    this.batch.del(key);
+  }
+
+  /**
+   * Commit the current batch.
+   * @returns {Promise}
+   */
+
+  async commit() {
+    await this.batch.write();
+    this.batch = null;
+  }
+
+  /**
+   * Open the indexer, open the database,
+   * initialize height, and bind to events.
+   * @returns {Promise}
+   */
+
+  async open() {
+    this.logger.info('Indexer is loading.');
+
+    await this.ensure();
+    await this.db.open();
+    await this.db.verify(layout.V.encode(), 'index', 0);
+    await this.verifyNetwork();
+
+    // Initialize the indexed height.
+    const data = await this.db.get(layout.R.encode());
+    if (data)
+      this.height = bio.readU32(data, 0);
+    else
+      await this.saveGenesis();
+
+    // Bind to chain events.
+    this.bind();
+  }
+
+  /**
+   * Close the indexdb, wait for the database to close.
+   * @returns {Promise}
+   */
+
+  async close() {
+    return this.db.close();
+  }
+
+  /**
+   * Ensure prefix directory (prefix/index).
+   * @returns {Promise}
+   */
+
+  async ensure() {
+    if (fs.unsupported)
+      return;
+
+    if (this.options.memory)
+      return;
+
+    await fs.mkdirp(this.options.prefix);
+  }
+
+  /**
+   * Verify network of index.
+   * @returns {Promise}
+   */
+
+  async verifyNetwork() {
+    let raw = await this.db.get(layout.O.encode());
+
+    if (!raw) {
+      raw = bio.write(4).writeU32(this.network.magic).render();
+      await this.db.put(layout.O.encode(), raw);
+      return;
+    }
+
+    const magic = bio.readU32(raw, 0);
+
+    if (magic !== this.network.magic)
+      throw new Error('Indexer: Network mismatch.');
+  }
+
+  /**
+   * A special case for indexing the genesis block. The genesis
+   * block coins are not spendable, however indexers can still index
+   * the block for historical and informational purposes.
+   * @private
+   * @returns {Promise}
+   */
+
+  async saveGenesis() {
+    this.start();
```

Comment:
We should move this `start` after `indexBlock` because most of the time `indexBlock` is going to have another `start` which can error with `Already started`.

Reply:
OK, I believe the intention was to have `indexBlock` assume a batch `start` and do the writes. :+1:

Reply:
Yep, `indexBlock` is adding to the batch writes.